// @solid/community-server/dist/server/notifications/Subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SUBSCRIBE_SCHEMA = void 0;
const iso8601_duration_1 = require("iso8601-duration");
const yup_1 = require("yup");
const Notification_1 = require("./Notification");
/**
 * A JSON parsing schema that can be used to parse subscription input.
 * Specific subscription types can extend this schema with their own custom keys.
 */
exports.SUBSCRIBE_SCHEMA = (0, yup_1.object)({
    '@context': (0, yup_1.array)((0, yup_1.string)()).ensure().required().test({
        name: 'RequireNotificationContext',
        message: `The ${Notification_1.CONTEXT_NOTIFICATION} context is required in the subscription JSON-LD body.`,
        test: (context) => Boolean(context?.includes(Notification_1.CONTEXT_NOTIFICATION)),
    }),
    type: (0, yup_1.string)().required(),
    topic: (0, yup_1.string)().required(),
    state: (0, yup_1.string)().optional(),
    expiration: (0, yup_1.number)().transform((value, original) => 
    // Convert the date string to milliseconds
    Date.parse(original)).optional(),
    rate: (0, yup_1.number)().transform((value, original) => 
    // Convert the rate string to milliseconds
    (0, iso8601_duration_1.toSeconds)((0, iso8601_duration_1.parse)(original)) * 1000).optional(),
    accept: (0, yup_1.string)().optional(),
});
