// clownface-shacl-path/index.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "findNodes", {
  enumerable: true,
  get: function () {
    return _findNodes.findNodes;
  }
});
Object.defineProperty(exports, "toSparql", {
  enumerable: true,
  get: function () {
    return _toSparql.toSparql;
  }
});
Object.defineProperty(exports, "PathVisitor", {
  enumerable: true,
  get: function () {
    return _path.PathVisitor;
  }
});
Object.defineProperty(exports, "ShaclPropertyPath", {
  enumerable: true,
  get: function () {
    return _path.ShaclPropertyPath;
  }
});
Object.defineProperty(exports, "ZeroOrMorePath", {
  enumerable: true,
  get: function () {
    return _path.ZeroOrMorePath;
  }
});
Object.defineProperty(exports, "OneOrMorePath", {
  enumerable: true,
  get: function () {
    return _path.OneOrMorePath;
  }
});
Object.defineProperty(exports, "PredicatePath", {
  enumerable: true,
  get: function () {
    return _path.PredicatePath;
  }
});
Object.defineProperty(exports, "AlternativePath", {
  enumerable: true,
  get: function () {
    return _path.AlternativePath;
  }
});
Object.defineProperty(exports, "InversePath", {
  enumerable: true,
  get: function () {
    return _path.InversePath;
  }
});
Object.defineProperty(exports, "SequencePath", {
  enumerable: true,
  get: function () {
    return _path.SequencePath;
  }
});
Object.defineProperty(exports, "ZeroOrOnePath", {
  enumerable: true,
  get: function () {
    return _path.ZeroOrOnePath;
  }
});
Object.defineProperty(exports, "fromNode", {
  enumerable: true,
  get: function () {
    return _path.fromNode;
  }
});

var _findNodes = require("./lib/findNodes.cjs");

var _toSparql = require("./lib/toSparql.cjs");

var _path = require("./lib/path.cjs");
