// @treecg/ldes-snapshot/dist/src/metadata/SnapshotMetadataInitializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotMetadataInitializer = void 0;
var SnapshotMetadata_1 = require("./SnapshotMetadata");
var Vocabularies_1 = require("../util/Vocabularies");
var SnapshotMetadataInitializer = /** @class */ (function () {
    function SnapshotMetadataInitializer() {
    }
    /**
     * Generate {@link ISnapshot} based on {@link ISnapshotOptions}.
     * @param options
     * @returns {SnapshotMetadata}
     */
    SnapshotMetadataInitializer.generateSnapshotMetadata = function (options) {
        var _a, _b, _c, _d;
        var date = (_a = options.date) !== null && _a !== void 0 ? _a : new Date();
        return new SnapshotMetadata_1.SnapshotMetadata({
            eventStreamIdentifier: (_b = options.snapshotIdentifier) !== null && _b !== void 0 ? _b : "http://example.org/" + date.valueOf(),
            materialized: false,
            members: [],
            snapshotOf: options.ldesIdentifier,
            snapshotUntil: date,
            timestampPath: (_c = options.timestampPath) !== null && _c !== void 0 ? _c : Vocabularies_1.DCT.created,
            versionOfPath: (_d = options.versionOfPath) !== null && _d !== void 0 ? _d : Vocabularies_1.DCT.isVersionOf
        });
    };
    return SnapshotMetadataInitializer;
}());
exports.SnapshotMetadataInitializer = SnapshotMetadataInitializer;
