// @treecg/ldes-snapshot/dist/src/metadata/SnapshotMetadata.js
"use strict";
/***************************************
 * Title: SnapshotMetadata
 * Description: Contains the interface and implementation for the metadata of a Snapshot (of a versioned LDES).
 * Author: Wout Slabbinck (wout.slabbinck@ugent.be)
 * Created on 30/11/2022
 *****************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnapshotMetadata = void 0;
var n3_1 = require("n3");
var Vocabularies_1 = require("../util/Vocabularies");
var TimestampUtil_1 = require("../util/TimestampUtil");
var namedNode = n3_1.DataFactory.namedNode;
var SnapshotMetadata = /** @class */ (function () {
    function SnapshotMetadata(args) {
        this.eventStreamIdentifier = args.eventStreamIdentifier;
        this.timestampPath = args.timestampPath;
        this.versionOfPath = args.versionOfPath;
        this.snapshotOf = args.snapshotOf;
        this.snapshotUntil = args.snapshotUntil;
        this.members = args.members;
        this.materialized = args.materialized;
    }
    SnapshotMetadata.prototype.getStore = function () {
        var _this = this;
        var store = new n3_1.Store();
        store.addQuad(namedNode(this.eventStreamIdentifier), Vocabularies_1.RDF.terms.type, Vocabularies_1.LDES.terms.EventStream);
        store.addQuad(namedNode(this.eventStreamIdentifier), Vocabularies_1.LDES.terms.timestampPath, namedNode(this.timestampPath));
        store.addQuad(namedNode(this.eventStreamIdentifier), Vocabularies_1.LDES.terms.versionOfPath, namedNode(this.versionOfPath));
        store.addQuad(namedNode(this.eventStreamIdentifier), Vocabularies_1.LDES.terms.snapshotOf, namedNode(this.snapshotOf));
        store.addQuad(namedNode(this.eventStreamIdentifier), Vocabularies_1.LDES.terms.snapshotUntil, (0, TimestampUtil_1.dateToLiteral)(this.snapshotUntil));
        this.members.forEach(function (snapshotMember) {
            store.addQuad(namedNode(_this.eventStreamIdentifier), Vocabularies_1.TREE.terms.member, namedNode(snapshotMember.id.value));
            store.addQuads(snapshotMember.quads);
        });
        return store;
    };
    return SnapshotMetadata;
}());
exports.SnapshotMetadata = SnapshotMetadata;
