// @solid/community-server/dist/authorization/AllStaticReader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AllStaticReader = void 0;
const IdentifierMap_1 = require("../util/map/IdentifierMap");
const PermissionReader_1 = require("./PermissionReader");
/**
 * PermissionReader which sets all permissions to true or false
 * independently of the identifier and requested permissions.
 */
class AllStaticReader extends PermissionReader_1.PermissionReader {
    constructor(allow) {
        super();
        this.permissions = Object.freeze({
            read: allow,
            write: allow,
            append: allow,
            create: allow,
            delete: allow,
        });
    }
    async handle({ credentials, requestedModes }) {
        const availablePermissions = new IdentifierMap_1.IdentifierMap();
        const permissions = this.createPermissions(credentials);
        for (const [identifier] of requestedModes) {
            availablePermissions.set(identifier, permissions);
        }
        return availablePermissions;
    }
    createPermissions(credentials) {
        const result = {};
        for (const group of Object.keys(credentials)) {
            result[group] = this.permissions;
        }
        return result;
    }
}
exports.AllStaticReader = AllStaticReader;
